File: robot/rospackages/src/mcu_control/scripts/APA102.py
def constrain(val, min_val, max_val):
  return min(max_val, max(min_val, val))


class APA102:
  def __init__(self, spi, num_of_leds):
    self.spi = spi
    self.num_of_leds = num_of_leds
    self.buf = []
    for i in range(4):
      self.buf.append(0x00)
    for i in range(4, num_of_leds + 4):
      self.buf.append(0xE0)
      for j in range(3):
        self.buf.append(0x00)
    size_end_frame = num_of_leds // 8 # size in bytes
    if size_end_frame * 8 < (num_of_leds / 2):
      size_end_frame += 1
    for i in range(size_end_frame):
      self.buf.append(0xFF)
  def send(self):
    self.spi(self.buf)
  def set_colour(self, number, r, g, b, brightness = 31):
    self.buf[4 * number + 4] = 0xE0 + constrain(brightness, 0, 31)
    self.buf[4 * number + 5] = constrain(b, 0, 255)
    self.buf[4 * number + 6] = constrain(g, 0, 255)
    self.buf[4 * number + 7] = constrain(r, 0, 255)
  def set_all(self, r, g, b, brightness = 31):
    for i in range(self.num_of_leds):
      self.set_colour(i, r, g, b, brightness)
    # print(self.buf)
  def set_to_off(self):
    self.set_all(0, 0, 0, 0)
  def set_pins(self, pins, r, g, b, brightness = 31):
    self.set_to_off()
    for i in pins:
      self.set_colour(i, r, g, b, brightness)

File: robot/rover/MobilePlatform/src/APA102.cpp
#include "APA102.h"

#define START_FRAME_SIZE 4

APA102::APA102(int numOfLEDs, int mosiPin, int clockPin)
                        : numOfLEDs(numOfLEDs), mosiPin(mosiPin), clockPin(clockPin) {
  int sizeEndFrame = numOfLEDs/16;
  if (sizeEndFrame * 8 < (numOfLEDs / 2)) {
    sizeEndFrame++;
  }
  bufSize = START_FRAME_SIZE + numOfLEDs * 4 + sizeEndFrame;
  buf = (byte*)malloc(bufSize);
  for (int i = 0; i < START_FRAME_SIZE; i++) {
    buf[i] = 0;
  }
  for (int i = START_FRAME_SIZE; i < 4 * numOfLEDs + START_FRAME_SIZE; i+= 4) {
    buf[i + 0] = 0xE0;
    buf[i + 1] = 0;
    buf[i + 2] = 0;
    buf[i + 3] = 0;
  }
  for (int i = START_FRAME_SIZE + numOfLEDs * 4; i < bufSize; i++) {
    buf[i] = 255;
  }  
}
void APA102::send() {
  for (int i = 0; i < bufSize; i++) {
    shiftOut(mosiPin, clockPin, MSBFIRST, buf[i]);
  }
}

void APA102::setColor(int index, int r, int g, int b, int brightness) {
  buf[4 * index + 4] = 0xE0 + constrain(brightness, 0, 31);
  buf[4 * index + 5] = constrain(b, 0, 255);
  buf[4 * index + 6] = constrain(g, 0, 255);
  buf[4 * index + 7] = constrain(r, 0, 255);
}

void APA102::setAll(int r, int g, int b, int brightness) {
  for (int i = 0; i < numOfLEDs; i++) {
    setColor(i, r, g, b, brightness);
  }
}

void APA102::setToOff() {
  setAll(0, 0, 0, 0);
}